.github/workflows/CD.yml

```diff
@@ -61,10 +65,69 @@ jobs:
           CONTAINER_REGISTRY_PASSWORD: ${{ secrets.CONTAINER_REGISTRY_PASSWORD }}
           CONTAINER_REGISTRY_USER: ${{ secrets.CONTAINER_REGISTRY_USER }}
         run: docker login quay.io --username $CONTAINER_REGISTRY_USER --password $CONTAINER_REGISTRY_PASSWORD
-      - name: Build and Publish JVM Container
+      - name: Build and Publish JVM Container - All in One
         shell: bash
         run: |
-          docker build -f runner/src/main/docker/Dockerfile.jvm -t $OUTPUT_CONTAINER_NAME-jvm runner/
-          docker push $OUTPUT_CONTAINER_NAME-jvm
-          docker image rm -f $OUTPUT_CONTAINER_NAME-jvm
-#
+          echo "Building and pushing docker image for All-in-One"
+          docker build -f docker/Dockerfile.jvm -t $ALL_IN_ONE_CONTAINER_NAME-jvm runner/
+          docker push $ALL_IN_ONE_CONTAINER_NAME-jvm
+          docker image rm -f $ALL_IN_ONE_CONTAINER_NAME-jvm
+
+      - uses: dorny/paths-filter@v2
+        id: filter
+        with:
+          filters: |
+            general &general:
+              - infra/**
+              - bom/**
+              - build-parent/**
+              - infra/**
+              - k8s/**
+            manager:
+              - *general
+              - manager/**
+            shard:
+              - *general
+              - shard/**
+            ingress:
+              - *general
+              - ingress/**
+            shard:
+              - *general
+              - executor/**
+
+      # run only if 'manager' or any 'general' modules have changed
+      - name: Build and Publish JVM Container - Fleet Manager
+        if: steps.filter.outputs.manager == 'true'
+        run: |
+          echo "Building and pushing docker image for Fleet Manager"
+          docker build -f docker/Dockerfile.jvm -t $FLEET_MANAGER_CONTAINER_NAME-jvm runner/
+          docker push $FLEET_MANAGER_CONTAINER_NAME-jvm
+          docker image rm -f $FLEET_MANAGER_CONTAINER_NAME-jvm
+
+      # run only if 'ingress' or any 'general' modules have changed
+      - name: Build and Publish JVM Container - Fleet Shard
+        if: steps.filter.outputs.shard == 'true'
+        run: |
+          echo "Building and pushing docker image for Fleet Shard"
+          docker build -f docker/Dockerfile.jvm -t $FLEET_SHARD_CONTAINER_NAME-jvm runner/
```

Comment:
@r00ta Do we need this `docker build` here? From what I see we're always building the same image no matter the component, and we always build the all-in-one first. We just need to retag (and push) the all-in-one image at this point.